src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+
+// Getter and setter for bool value voxel grids
+bool VoxelGrid::getBoolVoxelByIndex(const Mn::Vector3i& coords,
+                                    const std::string& gridName) {
+  assert(boolGrids_.find(gridName) != boolGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return boolGrids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setBoolVoxelByIndex(const Mn::Vector3i& coords,
+                                    bool val,
+                                    const std::string& gridName) {
+  assert(boolGrids_.find(gridName) != boolGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  boolGrids_[gridName].get()[hashedVoxelIndex] = val;
+}
+// Getter and setter for int value voxel grids
+int VoxelGrid::getIntVoxelByIndex(const Mn::Vector3i& coords,
+                                  const std::string& gridName) {
+  assert(intGrids_.find(gridName) != intGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return intGrids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setIntVoxelByIndex(const Mn::Vector3i& coords,
+                                   int val,
+                                   const std::string& gridName) {
+  assert(intGrids_.find(gridName) != intGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  intGrids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+// Getter and setter for Float value voxel grids
+float VoxelGrid::getFloatVoxelByIndex(const Mn::Vector3i& coords,
+                                      const std::string& gridName) {
+  assert(floatGrids_.find(gridName) != floatGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return floatGrids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setFloatVoxelByIndex(const Mn::Vector3i& coords,
+                                     float val,
+                                     const std::string& gridName) {
+  assert(floatGrids_.find(gridName) != floatGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  floatGrids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+// Getter and setter for Vector3 value voxel grids
+Mn::Vector3 VoxelGrid::getVector3VoxelByIndex(const Mn::Vector3i& coords,
+                                              const std::string& gridName) {
+  assert(vector3Grids_.find(gridName) != vector3Grids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return vector3Grids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setVector3VoxelByIndex(const Mn::Vector3i& coords,
+                                       const Mn::Vector3& val,
+                                       const std::string& gridName) {
+  assert(vector3Grids_.find(gridName) != vector3Grids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  vector3Grids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+std::shared_ptr<Mn::Trade::MeshData> VoxelGrid::getMeshData(
+    const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
```

Comment:
What happens if I get the mesh, change the underlying grid, then get the mesh again? Is there logic somewhere to ensure the mesh gets updated to reflect the changes to the underlying grid?

Reply:
I considered giving each mesh a boolean that would keep track of whether or not it's in sync with the underlying grid. The problem I think is that each time a voxel is changed, I'd have to set this boolean. Since a user could potentially set millions of voxels, I think it makes sense to leave it up to them to manually update the mesh (by calling generateMesh(gridName)) again when they need to. Does that answer your question, or do you think there is a better way?

Reply:
Manual/explicit calls to generateMesh sounds reasonable. As a nit, I'd prefer it to *always* be manual, even for the first generate, for the sake of consistency. As-is, you risk confusing a user who never discovered the "generateMesh" API and is surprised to be getting a stale mesh from getMeshData in some cases.

Reply:
I see - would a different naming scheme be more appropriate? I thought about doing updateMesh() instead, but the intitial creation of the mesh isn't updating anything since the mesh hasn't been made.